Based in patch written by:   Frank Schmirler <vdrdev@schmirler.de>
Project's homepage:          http://vdr.schmirler.de/

Latest version available at: http://vdr.schmirler.de/volctrl/
------------------------------------------------------------------------
This patch enables volume control with the left/right of your remote
control. Enabled it in OSD menu Setup > Miscellaneous.

Setup options:
- Channelgroups with left/right
Affects the behaviour of left/right while watching live TV.
  yes: Default VDR behaviour. No volume control with left/right
  only in channelinfo: While channelinfo is open, left/right will switch
       channelgroups (hit OK to display channelinfo if not visible).
       Otherwise left/right controls the volume (hit OK to close
       channelinfo if it is currently visible).
  no:  You will loose the channelgroup feature. Left/right will always
       control the volume
- Search fwd/back with left/right
Affects the behaviour of left/right during replay.
  yes: Default VDR behaviour. No volume control with left/right
  only in progress display: While progress bar is shown, left/right
       controls speed and direction (hit OK to display progress bar if
       not visible).
       Otherwise left/right controls the volume (hit OK to close
       progress bar if it is currently visible).
  no:  You will loose the ability to search with left/right. Makes sense
       if your remote control has dedicated keys for this.
------------------------------------------------------------------------

diff -Naur VDR-NG-orig/config.c VDR-NG/config.c
--- VDR-NG-orig/config.c	2009-06-12 18:10:14.000000000 +0200
+++ VDR-NG/config.c	2009-06-12 19:34:10.000000000 +0200
@@ -330,6 +330,8 @@
   ChannelFilter = 0;
   InitialChannel = 0;
   InitialVolume = -1;
+  LRChannelGroups = 0;
+  LRForwardRewind = 0;
   ShowRecDate = 1;
   ShowRecTime = 1;
   ShowRecLength = 1;
@@ -524,6 +526,8 @@
   else if (!strcasecmp(Name, "ChannelFilter"))       ChannelFilter      = atoi(Value);
   else if (!strcasecmp(Name, "InitialChannel"))      InitialChannel     = atoi(Value);
   else if (!strcasecmp(Name, "InitialVolume"))       InitialVolume      = atoi(Value);
+  else if (!strcasecmp(Name, "LRChannelGroups"))     LRChannelGroups    = atoi(Value);
+  else if (!strcasecmp(Name, "LRForwardRewind"))     LRForwardRewind    = atoi(Value);
   else if (!strcasecmp(Name, "ShowRecDate"))         ShowRecDate        = atoi(Value);
   else if (!strcasecmp(Name, "ShowRecTime"))         ShowRecTime        = atoi(Value);
   else if (!strcasecmp(Name, "ShowRecLength"))       ShowRecLength      = atoi(Value);
@@ -625,6 +629,8 @@
   Store("ChannelFilter",      ChannelFilter);
   Store("InitialChannel",     InitialChannel);
   Store("InitialVolume",      InitialVolume);
+  Store("LRChannelGroups",    LRChannelGroups);
+  Store("LRForwardRewind",    LRForwardRewind);
   Store("ShowRecDate",        ShowRecDate);
   Store("ShowRecTime",        ShowRecTime);
   Store("ShowRecLength",      ShowRecLength);
diff -Naur VDR-NG-orig/config.h VDR-NG/config.h
--- VDR-NG-orig/config.h	2009-06-12 18:10:14.000000000 +0200
+++ VDR-NG/config.h	2009-06-12 19:34:10.000000000 +0200
@@ -288,6 +288,8 @@
   int ChannelFilter;
   int InitialChannel;
   int InitialVolume;
+  int LRChannelGroups;
+  int LRForwardRewind;
   int ShowRecDate, ShowRecTime, ShowRecLength, ShowProgressBar, MenuCmdPosition;
   int RecordingsSortMode;
   int RecordingsSortDirsFirst;
diff -Naur VDR-NG-orig/i18n.c VDR-NG/i18n.c
--- VDR-NG-orig/i18n.c	2009-06-12 18:10:14.000000000 +0200
+++ VDR-NG/i18n.c	2009-06-12 19:41:40.000000000 +0200
@@ -5450,6 +5450,94 @@
     "",// TODO
     "",// hrv TODO
   },
+  { "Setup.Miscellaneous$Channelgroups with left/right",
+    "Kanalgruppen mit Rechts/Links",
+    "",// TODO
+    "",// TODO
+    "",// TODO
+    "",// TODO
+    "",// TODO
+    "",// TODO
+    "",// TODO
+    "",// TODO
+    "Grupos de canales con izq./dcha.",
+    "",// TODO
+    "",// TODO
+    "",// TODO
+    "",// TODO
+    "",// TODO
+    "",// TODO
+    "",// TODO
+    "",// TODO
+    "",// TODO
+    "",// TODO
+  },
+  { "Setup.Miscellaneous$only in channelinfo",
+    "nur in Kanalinfo",
+    "",// TODO
+    "",// TODO
+    "",// TODO
+    "",// TODO
+    "",// TODO
+    "",// TODO
+    "",// TODO
+    "",// TODO
+    "en info de canal",
+    "",// TODO
+    "",// TODO
+    "",// TODO
+    "",// TODO
+    "",// TODO
+    "",// TODO
+    "",// TODO
+    "",// TODO
+    "",// TODO
+    "",// TODO
+  },
+  { "Setup.Miscellaneous$Search fwd/back with left/right",
+    "Vor-/Rücklauf mit Rechts/Links",
+    "",// TODO
+    "",// TODO
+    "",// TODO
+    "",// TODO
+    "",// TODO
+    "",// TODO
+    "",// TODO
+    "",// TODO
+    "Rebobinar/Avanzar con izq./dcha.",
+    "",// TODO
+    "",// TODO
+    "",// TODO
+    "",// TODO
+    "",// TODO
+    "",// TODO
+    "",// TODO
+    "",// TODO
+    "",// TODO
+    "",// TODO
+  },
+  { "Setup.Miscellaneous$only in progress display",
+    "nur in Fortschrittsanzeige",
+    "",// TODO
+    "",// TODO
+    "",// TODO
+    "",// TODO
+    "",// TODO
+    "",// TODO
+    "",// TODO
+    "",// TODO
+    "en barra de reproducción",
+    "",// TODO
+    "",// TODO
+    "",// TODO
+    "",// TODO
+    "",// TODO
+    "",// TODO
+    "",// TODO
+    "",// TODO
+    "",// TODO
+    "",// TODO
+  },
   // The days of the week:
   { "MTWTFSS",
     "MDMDFSS",
diff -Naur VDR-NG-orig/menu.c VDR-NG/menu.c
--- VDR-NG-orig/menu.c	2009-06-12 18:10:14.000000000 +0200
+++ VDR-NG/menu.c	2009-06-12 19:34:10.000000000 +0200
@@ -3258,6 +3258,8 @@
 class cMenuSetupMisc : public cMenuSetupBase {
 private:
   const char *channelFilterTexts[3];
+  const char *lrChannelGroupsTexts[3];
+  const char *lrForwardRewindTexts[3];
 public:
   cMenuSetupMisc(void);
   };
@@ -3268,6 +3270,12 @@
   channelFilterTexts[0] = tr("no");
   channelFilterTexts[1] = tr("Tv only");
   channelFilterTexts[2] = tr("Radio only");
+  lrChannelGroupsTexts[0] = tr("no");
+  lrChannelGroupsTexts[1] = tr("Setup.Miscellaneous$only in channelinfo");
+  lrChannelGroupsTexts[2] = tr("yes");
+  lrForwardRewindTexts[0] = tr("no");
+  lrForwardRewindTexts[1] = tr("Setup.Miscellaneous$only in progress display");
+  lrForwardRewindTexts[2] = tr("yes");
   Add(new cMenuEditIntItem( tr("Setup.Miscellaneous$Min. event timeout (min)"),   &data.MinEventTimeout));
   Add(new cMenuEditIntItem( tr("Setup.Miscellaneous$Min. user inactivity (min)"), &data.MinUserInactivity));
   Add(new cMenuEditIntItem( tr("Setup.Miscellaneous$SVDRP timeout (s)"),          &data.SVDRPTimeout));
@@ -3275,6 +3283,8 @@
   Add(new cMenuEditStraItem(tr("Setup.Miscellaneous$Filter channels"),            &data.ChannelFilter, 3, channelFilterTexts));;    
   Add(new cMenuEditChanItem(tr("Setup.Miscellaneous$Initial channel"),            &data.InitialChannel, tr("Setup.Miscellaneous$as before")));
   Add(new cMenuEditIntItem( tr("Setup.Miscellaneous$Initial volume"),             &data.InitialVolume, -1, 255, tr("Setup.Miscellaneous$as before")));
+  Add(new cMenuEditStraItem(tr("Setup.Miscellaneous$Channelgroups with left/right"),   &data.LRChannelGroups, 3, lrChannelGroupsTexts));
+  Add(new cMenuEditStraItem(tr("Setup.Miscellaneous$Search fwd/back with left/right"), &data.LRForwardRewind, 3, lrForwardRewindTexts));
   Add(new cMenuEditBoolItem(tr("Setup.Miscellaneous$Show cancel message at shutdown"), &data.ShutdownMessage, tr("no"), tr("yes")));
   Add(new cMenuEditBoolItem(tr("Setup.Miscellaneous$Abort when Plugin fails to load"), &data.AbortWhenPluginFails));
   Add(new cMenuEditBoolItem(tr("Setup.Miscellaneous$Emergency exit"),             &data.EmergencyExit));
@@ -4055,6 +4065,11 @@
     case kLeft:
     case kRight|k_Repeat:
     case kRight:
+         if (!Setup.LRChannelGroups) {
+	    cRemote::Put(NORMALKEY(Key) == kLeft ? kVolDn : kVolUp, true);
+	    break;
+            }
+         // else fall through
     case kNext|k_Repeat:
     case kNext:
     case kPrev|k_Repeat:
@@ -4236,18 +4251,13 @@
 eOSState cDisplayVolume::ProcessKey(eKeys Key)
 {
   switch (Key) {
-//M7X0 BEGIN AK
-// Support Volume Change via Left/Right added
-//M7X0TODO: Make this working in replaying mode as well
-		case kLeft|k_Repeat:
-		case kLeft:
-		case kRight|k_Repeat:
-		case kRight:
-			cDevice::PrimaryDevice()->SetVolume(NORMALKEY(Key) == kLeft ? -VOLUMEDELTA : VOLUMEDELTA);
-			Show();
-			timeout.Set(VOLUMETIMEOUT);
-			break;
-//M7X0 END AK
+    case kLeft|k_Repeat:
+    case kLeft:
+    case kRight|k_Repeat:
+    case kRight:
+         cRemote::Put(NORMALKEY(Key) == kLeft ? kVolDn : kVolUp, true);
+         break;
+         // else fall through
     case kVolUp|k_Repeat:
     case kVolUp:
     case kVolDn|k_Repeat:
@@ -5125,6 +5135,20 @@
      return osContinue;
      }
   bool DoShowMode = true;
+  if (!Setup.LRForwardRewind || (Setup.LRForwardRewind == 1 && !visible)) {
+    switch (Key) {
+      // Left/Right volume control
+      case kLeft|k_Repeat:
+      case kLeft:
+      case kRight|k_Repeat:
+      case kRight:
+	  cRemote::Put(NORMALKEY(Key) == kLeft ? kVolDn : kVolUp, true);
+        return osContinue;
+        break;
+      default:
+        break;
+    }
+  }
   switch (Key) {
     // Positioning:
     case kPlay:
diff -Naur VDR-NG-orig/vdr.c VDR-NG/vdr.c
--- VDR-NG-orig/vdr.c	2009-06-12 18:10:14.000000000 +0200
+++ VDR-NG/vdr.c	2009-06-12 20:01:22.000000000 +0200
@@ -1093,38 +1093,19 @@
           case kVolDn|k_Repeat:
           case kVolDn:
           case kMute:
-//M7X0 BEGIN AK
-// Taken from gambler
-          // m7x0 volume control via left/right
-          case kLeft|k_Repeat:
-          case kLeft:
-          case kRight|k_Repeat:
-          case kRight:
                if (key == kMute) {
                   if (!cDevice::PrimaryDevice()->ToggleMute() && !Menu) {
                      key = kNone; // nobody else needs to see these keys
                      break; // no need to display "mute off"
                      }
-
-                  if (!Menu && !cOsd::IsOpen())
-                     Menu = cDisplayVolume::Create();
-
-                  cDisplayVolume::Process(key);
-                  key = kNone;
-                  }
-               else if ((!Menu && !cOsd::IsOpen()) || NORMALKEY(key) == kVolUp  || NORMALKEY(key) == kVolDn) {
-                  if (NORMALKEY(key) == kVolUp  || NORMALKEY(key) == kVolDn)
-                     cDevice::PrimaryDevice()->SetVolume(NORMALKEY(key) == kVolDn ? -VOLUMEDELTA : VOLUMEDELTA);
-                  if (!Menu && !cOsd::IsOpen())
-                     Menu = cDisplayVolume::Create();
-
-                  cDisplayVolume::Process(key);
-                  key = kNone; // nobody else needs to see these keys
-                               // m7x0 we need the l/r keys in setup menu,
-                               // keep it in menu mode
                   }
+               else
+                  cDevice::PrimaryDevice()->SetVolume(NORMALKEY(key) == kVolDn ? -VOLUMEDELTA : VOLUMEDELTA);
+               if (!Menu && !cOsd::IsOpen())
+                  Menu = cDisplayVolume::Create();
+               cDisplayVolume::Process(key);
+               key = kNone; // nobody else needs to see these keys
                break;
-//M7X0 END AK
           // Audio track control:
           case kAudio:
                if (cControl::Control())
@@ -1274,18 +1255,23 @@
                   Channels.SwitchTo(PreviousChannel[PreviousChannelIndex ^= 1]);
                   break;
                   }
+             // Left/Right volume control
+             case kLeft|k_Repeat:
+             case kLeft:
+             case kRight|k_Repeat:
+             case kRight:
+		  if (Setup.LRChannelGroups < 2) {
+                     cRemote::Put(NORMALKEY(key) == kLeft ? kVolDn : kVolUp, true);
+                     break;
+                     }
+                  // else fall through
              // Direct Channel Select:
              case k1 ... k9:
-//M7X0 BEGIN AK
              // Previous/Next rotates through channel groups:
              case kPrev|k_Repeat:
              case kPrev:
              case kNext|k_Repeat:
              case kNext:
-             //case kLeft|k_Repeat:
-             //case kLeft:
-             //case kRight|k_Repeat:
-             //case kRight:
              // Up/Down Channel Select:
              case kUp|k_Repeat:
              case kUp:
@@ -1293,7 +1279,6 @@
              case kDown:
                   Menu = new cDisplayChannel(NORMALKEY(key));
                   break;
-//M7X0 END AK
              // Viewing Control:
              case kOk:   LastChannel = -1; break; // forces channel display
              // Instant resume of the last viewed recording:
